File: problems/0015-3sum/solution.py
"""
Problem Number: 15
Problem Name: 3Sum
Difficulty: Medium
Topics: Array, Two Pointers, Sorting
Companies: Facebook, Amazon, Microsoft
"""

from typing import List

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        """
        Approach: Sorting + Two Pointers
        Time Complexity: O(n²)
        Space Complexity: O(1) (ignoring output)
        """
        res = []
        nums.sort()
        
        for i in range(len(nums) - 2):
            if i > 0 and nums[i] == nums[i-1]:
                continue
                
            l, r = i + 1, len(nums) - 1
            while l < r:
                s = nums[i] + nums[l] + nums[r]
                if s < 0:
                    l += 1
                elif s > 0:
                    r -= 1
                else:
                    res.append([nums[i], nums[l], nums[r]])
                    while l < r and nums[l] == nums[l+1]:
                        l += 1
                    while l < r and nums[r] == nums[r-1]:
                        r -= 1
                    l += 1
                    r -= 1
        return res

# Test cases
if __name__ == "__main__":
    solution = Solution()
    print(solution.threeSum([-1,0,1,2,-1,-4])) # [[-1,-1,2],[-1,0,1]]
    print(solution.threeSum([0,1,1]))          # []
    print(solution.threeSum([0,0,0]))          # [[0,0,0]]

File: problems/0015-3sum/solution.js
/**
 * @problemNumber: 15
 * @problemName: 3Sum
 * @difficulty: Medium
 * @topics: Array, Two Pointers, Sorting
 * @companies: Facebook, Amazon, Microsoft
 */

/**
 * Approach: Sorting + Two Pointers
 * Time Complexity: O(n²)
 * Space Complexity: O(log n) to O(n) (sorting)
 *
 * @param {number[]} nums
 * @return {number[][]}
 */
function threeSum(nums) {
    const result = [];
    nums.sort((a, b) => a - b);

    for (let i = 0; i < nums.length - 2; i++) {
        // Skip duplicates for the first element
        if (i > 0 && nums[i] === nums[i - 1]) continue;

        let left = i + 1;
        let right = nums.length - 1;

        while (left < right) {
            const sum = nums[i] + nums[left] + nums[right];

            if (sum === 0) {
                result.push([nums[i], nums[left], nums[right]]);
                // Skip duplicates for the second element
                while (left < right && nums[left] === nums[left + 1]) left++;
                // Skip duplicates for the third element
                while (left < right && nums[right] === nums[right - 1]) right--;
                left++;
                right--;
            } else if (sum < 0) {
                left++;
            } else {
                right--;
            }
        }
    }

    return result;
}

// Test cases
console.log(threeSum([-1, 0, 1, 2, -1, -4])); // [[-1,-1,2],[-1,0,1]]
console.log(threeSum([0, 1, 1]));          // []
console.log(threeSum([0, 0, 0]));          // [[0,0,0]]

module.exports = { threeSum };
